code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+
+It is possible to combine type hints and initialization on the same line, like so:
+
+
+    foo: int = 3
+
+
+That looks somewhat redundant, doesn't it? How can the *literal* `3` be anything else than an integer? This is where type inference kicks in. Tools such as `mypy` will try to infer the types of variables from their use. It is quite safe to assume type inference is possible here, so probably best to just write:
+
+
+    foo = 3
+
+
+Type inference does have its limitations, for instance `mypy` will not do any type inference in functions without type hints. To understand why, let's quickly look into function type hints. In the simplest form:
+
+
+    def add(a: int, b: int) -> int:
+        c = a + b
+        return c
+
+
+The syntax is relatively straight forward, using the colon (`:`) for parameter type hints, and the arrow (`->`) for the return type. Notice how the type of `c` is not annotated. It can be, but it is not needed. From the types of `a` and `b` and the `+` operation, `mypy` can infer the type of `c`. But what if we did not annotate this function. Well, in that case, `a` and `b` could be anything: `str`, `float`, `list`, you name it! This is where `mypy` draws a line, if you do not annotate a function, `mypy` will not even attempt to do type inference. Instead all variables will be of type `Any`.
+
+What is `Any`? Well, anything really. It is an escape hatch of sorts that provides no information. Once `Any` gets involved type checking becomes rather impossible. What is `Any + int`? `Any`
+
+</details>
+
+1. Annotate the `factorial` function below:
+
+
+        def factorial(num):
+            total = 1
+            for i in range(2, num + 1):
+                total *= i
+            return total
+
+
+    <textarea name="form[q1]" rows="5" required=""></textarea>
+
+<details markdown="1"><summary  markdown="span">Generics</summary>
+
+Integers, floats, booleans and strings are primitive data types. Built into the language, they serve as building blocks for more complex data structures. For instance, you might need a `list` to store your data. 
+
+
+    numbers: list = [1, 2, 3]
+    number = numbers.pop()
+
+
+Here is the catch, the type `list` does not tell *anything* about what is in the `list`. So really what we have here is a `list` containing `Any`. In this case the type of `number` would be `Any` too.
+
+A `list` is a generic data type. It can store various types, but its operation will vary based on what you store. Simply put for a `list`, if you initially store integers in the list, you will later be able to retrieve integers from that list. This can be annotated as follows:
+
+
+    numbers: list[int] = [1, 2, 3]
+    number = numbers.pop()
+
+
+Now `numbers` is defined as a list of integers, and through that `number` will be of type `int` too.
+
+Let's take a quick look at `dict`. Dictionaries are generic over two types, their keys and values. This is how that can be annotated:
```

Comment:
"are generic over" is een taalconstructie die ik in ieder geval niet terugvind in https://www.python.org/dev/peps/pep-0585/

Misschien een beetje te abstracte taal.